infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/ClusterDescriptor.java

```diff
@@ -16,67 +14,36 @@ public class ClusterDescriptor {
     private static int CORFU_PORT = 9000;
 
     @Getter
-    String clusterId;
+    final String clusterId;
 
     @Getter
-    ClusterRole role;
+    final ClusterRole role;
 
     @Getter
-    List<NodeDescriptor> nodesDescriptors;
+    final List<NodeDescriptor> nodeDescriptors;
 
     @Getter
-    private int corfuPort;    // Port on which Corfu DB runs on this cluster
+    final private int corfuPort;    // Port on which Corfu DB runs on this cluster
 
-    public ClusterDescriptor(String clusterId) {
-        this.clusterId = clusterId;
-    }
-
-    public ClusterDescriptor(ClusterConfigurationMsg clusterConfig) {
-        this.clusterId = clusterConfig.getId();
-        this.role = clusterConfig.getRole();
-        this.corfuPort = clusterConfig.getCorfuPort() != 0 ? clusterConfig.getCorfuPort() : CORFU_PORT;
-        this.nodesDescriptors = new ArrayList<>();
-        for (NodeConfigurationMsg nodeConfig : clusterConfig.getNodeInfoList()) {
-            NodeDescriptor newNode = new NodeDescriptor(nodeConfig.getAddress(),
-                    Integer.toString(nodeConfig.getPort()), clusterId,
-                    nodeConfig.getConnectionId(),
-                    nodeConfig.getNodeId());
-            this.nodesDescriptors.add(newNode);
-        }
-    }
 
     public ClusterDescriptor(ClusterDescriptor info, ClusterRole roleType) {
         this.clusterId = info.clusterId;
         this.role = roleType;
-        this.nodesDescriptors = new ArrayList<>();
+        this.nodeDescriptors = new ArrayList<>();
         this.corfuPort = info.getCorfuPort();
-        for (NodeDescriptor nodeInfo : info.nodesDescriptors) {
+        for (NodeDescriptor nodeInfo : info.nodeDescriptors) {
             NodeDescriptor newNode = new NodeDescriptor(nodeInfo.getHost(), nodeInfo.getPort(),
                     info.clusterId, nodeInfo.getConnectionId(), nodeInfo.getNodeId());
-            this.nodesDescriptors.add(newNode);
+            this.nodeDescriptors.add(newNode);
```

Comment:
I see that instead of having 1 representation of a cluster, you have multiple per model.. Still couple of questions/comments here:

(1) It's still loosely coupled to the replication model, so you have the same cluster with different object representations, something like: o1->role:source, o2->role:sink, o3->role:source, where 01, 02 and 02 actually refer to the same cluster and yet it has no info about to which model its a source or a sink. So you'll need a additional level of indirection.

(2) If you decide to go with this definition perhaps a clone method would make it cleaner. Also I'd like to know which advantage it does have.

(3) My suggestion, topologyDescriptor has a field Map<Model, Role> .. so the same cluster can have different roles according to the model which is encoded by the adapter/plugin. Example, a global node for my customer is a SOURCE for model FULL_TABLE. and since this cluster is global, I'll set it in such a way. 

Note: here we are not covering can a same node have different roles for the same model according to the client? We don't have any such use case, and given that we don't have clients registering in advance I'd say either, we discard this option or, we allow it to be forced, example:

Map<Model, <role, client>>... if client field is empty it applies to all clients, if a specific client name is found it indicates It has this behavior for that client whenever in time it registers.

Reply:
will sync up offline about this.